### PR TITLE
Make adjustments per the OT-go SpanContext changes

### PR DESCRIPTION
**Pull request by bensigelman**
Make the zipkin-go-opentracing Context object implement opentracing.SpanContext, migrate the old Join concept to Extract, fix per the StartSpan changes, and get all tests to pass.

I am confident about the OpenTracing pieces of this but will point out a few Zipkin-related questions I have.

cc: @basvanbeek

### CODE REVIEW

File: context.go [hunk 1]
```diff
@@ -1,7 +1,16 @@
 package zipkintracer
 
-// Context holds the basic Span metadata.
+import (
+	"sync"
+
+	"github.com/opentracing/opentracing-go"
+)
+
+// Context holds the BasicSpan metadata that propagates across process
```
Comment (bensigelman) [inline context.go]: (I did some light updating of comments)

File: event.go [hunk 2]
```diff
@@ -14,11 +14,6 @@ type EventTag struct {
 	Value interface{}
 }
 
-// EventBaggage is received when SetBaggageItem is called.
-type EventBaggage struct {
```
Comment (bensigelman) [inline event.go]: I got rid of this since it was annoying to implement... if you care, it'd be easy enough to add back.

File: propagation.go [hunk 3]
```diff
@@ -21,63 +17,49 @@ type DelegatingCarrier interface {
...
-		sp.raw.Baggage[k] = v
+		sc.Baggage[k] = v
 	})
 
-	traceID, spanID, parentSpanID, sampled := ac.State()
-	sp.raw.Context = Context{
-		TraceID: traceID,
-		Sampled: sampled,
-	}
-	if p.tracer.options.clientServerSameSpan {
```
Comment (bensigelman) [inline propagation.go]: as evidence that this SpanContext change is a good thing, I will mention that the subtle `clientServerSameSpan` logic has been consolidated a bit.

File: span.go [hunk 4]
```diff
@@ -145,57 +137,14 @@ func (s *spanImpl) FinishWithOptions(opts opentracing.FinishOptions) {
 		// operation since s.tracer is accessed on every call to `Lock`.
 		s.reset()
 	}
-	spanPool.Put(s)
-}
-
-func (s *spanImpl) SetBaggageItem(restrictedKey, val string) opentracing.Span {
```
Comment (bensigelman) [inline span.go]: (moved to SpanContext)

File: tracer.go [hunk 5]
```diff
@@ -187,35 +185,54 @@ func (t *tracerImpl) StartSpanWithOptions(
...
-	// a opentracing.Span.
-	sp := t.getSpan()
-	if opts.Parent == nil {
-		sp.raw.TraceID, sp.raw.SpanID = randomID2()
-		sp.raw.Sampled = t.options.shouldSample(sp.raw.TraceID)
-	} else {
-		pr := opts.Parent.(*spanImpl)
-		sp.raw.TraceID = pr.raw.TraceID
-		sp.raw.SpanID = randomID()
-		if t.options.clientServerSameSpan {
```
Comment (bensigelman) [inline tracer.go]: I did not understand why this conditional needed to be here, so I removed it... seems to me that we should only care about `clientServerSameSpan` for RPCs, and we handle that case just below.

File: propagation_test.go [hunk 6]
```diff
@@ -83,7 +84,7 @@ func TestSpanPropagator(t *testing.T) {
 	exp.Start = time.Time{}.Add(1)
 
 	for i, sp := range spans {
-		if a, e := sp.ParentSpanID, exp.ParentSpanID; a != e {
+		if a, e := sp.ParentSpanID, exp.SpanID; a != e {
```
Comment (bensigelman) [inline propagation_test.go]: I thought this should be `exp.SpanID`... it's also the value that gets the test to pass. :)  Let me know if I misunderstand something about Zipkin here.

**Comment by bensigelman** [comment]
https://github.com/opentracing/opentracing-go/pull/82 is in which will break `zipkin-go-opentracing` until this PR is merged...

**Comment by bensigelman** [comment]
@basvanbeek the `basictracer-go` change is merged now, FYI.

**Comment by bensigelman** [comment]
the merge conflicts are non-trivial in that there's a lot of very zipkin-specific flag logic... let me know how you'd like to proceed. I did fix the implicitly copied `Mutex` that I also had to fix in `basictracer-go` (from which this was forked, yes?).

**Comment by basvanbeek** [comment]
that is correct... i will look into the flag logic tomorrow morning when I'm fresh... most of it has to do with binary fixed width serialization and omission of features in certain Zipkin libraries when trying to interop. lots of legacy to deal with

**Comment by bensigelman** [comment]
@basvanbeek makes sense. I am happy to help on the OT front but am concerned that I might inadvertently break stuff.

Really, the change I made is straightforward. Essentially:
- Move Baggage into the `Context` struct
- `s/Join/Extract/` and simplify in the process since it no longer needs to create Spans or deal with parentage... it's just the inverse of `Inject` now (yay, simpler)
- Protect the Baggage with its own lock and make `Context` a pointer within `RawSpan` (sigh); which conceptually makes sense because it is allowed to outlive the Span.
- ... then deal with compiler errors and other misc fallout

**Comment by basvanbeek** [comment]
since I had to rework some items I have manually merged your commits and updated some items. Thanks for helping out!

**Pull request closed without merging**